### PR TITLE
Update Microsoft Photos Exporter

### PR DESCRIPTION
**Pull request by gpriess**
Primarily, this change to MicrosoftPhotosExporter allows albums and photos to be exported by pagination size (to help prevent memory pressure during export). Previously, folders were recursively searched and photos data was populated and cached during one export operation. Now, the page size (set to 50 items), is used to limit the number of photos and albums exported in one operation. Also, the MicrosoftPhotosExporter supports access token refresh.

It also contains the following bug fixes:
- Importing albumless photos in MicrosoftPhotosImporter now works
- Sets the 'title' field in GooglePhotosExporter (required for MicrosoftPhotosImporter to create a new item by filename)
- MicrosoftPhotosImporter allows photos to get uploaded with the same filename by renaming the duplicate

### CODE REVIEW

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/mediaModels/GoogleMediaItem.java [hunk 1]
```diff
@@ -18,9 +18,7 @@
 
 import com.fasterxml.jackson.annotation.JsonProperty;
 
-/**
- Media item returned by queries to the Google Photos API.  Represents what is stored by Google.
- */
+/** Media item returned by queries to the Google Photos API. Represents what is stored by Google. */
```
Comment (wmorland) [inline extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/mediaModels/GoogleMediaItem.java]: Can you please reverse these formatting changes on the pull request. Unfortunately it makes the pull request harder to review and also pollutes the git history which makes tools like blame less helpful.

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/mediaModels/GoogleMediaItem.java [hunk 2]
```diff
@@ -42,28 +40,56 @@
...
+    this.id = id;
+  }
 
-  public void setId(String id) { this.id = id; }
+  public void setMimeType(String mimeType) {
+    this.mimeType = mimeType;
+  }
 
-  public void setMimeType(String mimeType) { this.mimeType = mimeType; }
+  public void setFilename(String filename) {
```
Comment (wmorland) [inline extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/mediaModels/GoogleMediaItem.java]: This is an example of the difficulty it causes. This setFilename method has been added but it is hard to spot amongst all the formatting changes.

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/mediaModels/GoogleMediaItem.java (hunk 1 above)
Comment (wmorland) [reply extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/mediaModels/GoogleMediaItem.java]: Agreed, this makes sense. We may want to add a step to the Travis CI to check this. I have created https://github.com/google/data-transfer-project/issues/739 to track this

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/datatransferproject/transfer/microsoft/MicrosoftTransferExtension.java [hunk 3]
```diff
@@ -48,51 +62,24 @@ public String getServiceId() {
...
-      return new MicrosoftContactsExporter(BASE_GRAPH_URL, client, mapper, transformerService);
-    }
-    if (transferDataType.equals(CALENDAR)) {
-      return new MicrosoftCalendarExporter(BASE_GRAPH_URL, client, mapper, transformerService);
-    }
-
-    if (transferDataType.equals(PHOTOS)) {
-      return new MicrosoftPhotosExporter(BASE_GRAPH_URL, client, mapper, jobStore);
-    }
+    Preconditions.checkArgument(initialized);
```
Comment (wmorland) [inline extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/datatransferproject/transfer/microsoft/MicrosoftTransferExtension.java]: I believe checkState is more appropriate as this is not dependent on argument https://github.com/google/guava/wiki/PreconditionsExplained

```suggestion
    Preconditions.checkState(initialized);
```

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/datatransferproject/transfer/microsoft/MicrosoftTransferExtension.java [hunk 4]
```diff
@@ -48,51 +62,24 @@ public String getServiceId() {
...
-      return new MicrosoftOfflineDataExporter(BASE_GRAPH_URL, client, mapper);
+    if (!offlineData && transferDataType.equals(OFFLINE_DATA)) {
+      // only enable if derived data explicitly set as a configuration value
+      // TODO we may want to provide a config option that allows deployers to disable transfer of
+      // certain data types
+      return null;
     }
 
-    return null;
+    Preconditions.checkArgument(SUPPORTED_EXPORT_SERVICES.contains(transferDataType));
```
Comment (wmorland) [inline extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/datatransferproject/transfer/microsoft/MicrosoftTransferExtension.java]: Here `checkArgument` seems appropriate

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/datatransferproject/transfer/microsoft/MicrosoftTransferExtension.java [hunk 5]
```diff
@@ -48,51 +62,24 @@ public String getServiceId() {
...
-      return new MicrosoftCalendarImporter(
-          BASE_GRAPH_URL, client, mapper, transformerService);
-    }
-
-    if (transferDataType.equals(PHOTOS)) {
-      return new MicrosoftPhotosImporter(BASE_GRAPH_URL, client, mapper, jobStore, monitor);
-    }
-
-    return null;
+    Preconditions.checkArgument(initialized);
```
Comment (wmorland) [inline extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/datatransferproject/transfer/microsoft/MicrosoftTransferExtension.java]: `checkState`

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/datatransferproject/transfer/microsoft/MicrosoftTransferExtension.java (hunk 3 above)
Comment (gpriess) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/datatransferproject/transfer/microsoft/MicrosoftTransferExtension.java]: Done!

File: extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/datatransferproject/transfer/microsoft/MicrosoftTransferExtension.java (hunk 5 above)
Comment (gpriess) [reply extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/datatransferproject/transfer/microsoft/MicrosoftTransferExtension.java]: Done!

**Comment by gpriess** [comment]
@wmorland thanks for taking the time to review this!

**Pull request merged**